mitodl/mitxonline
# Adds flexible pricing form

#### Pre-Flight checklist


#### What are the relevant tickets?
Closes #573

#### What's this PR do?
Implements #573: Adds a Flexible Pricing Request Form to the available Wagtail forms.

Within the CMS, the form has customizable blocks for intro text and post-submit text, text for guest users, and text for  application in progress, application approved, and application denied states. Admins can customize the form fields, but the form must contain a Your Income and Income Currency field to capture the user's yearly income (similar to the form on MicroMasters). Upon submission, the form submission also creates a FlexiblePrice record, which can be used to track processing of the request. 

Applicants must be logged in to access the form - otherwise, the Guest User text is displayed and the user cannot submit an application. They must not have a prior submission. Applicants with a submission will see the appropriate text for the status of their submission (in progress, approved, skipped). 

#### How should this be manually tested?
Log in as an administrator and add Currency Exchange Rates via the Django admin. You'll need at least one (an easy one is USD since that's always 1-to-1). The Description field will be displayed in the front end if supplied, so, for completeness, at least two should be supplied, and one should not have a description.

Log in as an administrator and navigate to the CMS admin. Create a new child page under a given course that's in there - you should be given the Flexible Pricing Request Form. Supply the necessary data. You will need to create at least two form fields: one named Your Income and one named Income Currency. The "Your Income" field should be a number field, and the "Income Currency" should be a Country field. After saving the form, go back to the course page in the CMS and add a link to the form in the content (this can be in any part of the course page content that'll render a link, it doesn't have to be in a specific section). 

Then, navigate to the course page and to the form. You should see one of three things, depending on state:
1. _If you're not logged in:_ You should see whatever you put in for Guest Text, and no form.
2. _If you're logged in and haven't submitted the form:_ You should see the form.
3. _If you're logged in and have submitted the form:_ You should see one of Application In Progress, Application Approved, or Application Denied texts. 

If you've submitted an application, you should:
1. See the form submission in the Wagtail admin (under Forms)
2. See the form submission in Django Admin, in FlexiblePricing under submissions. (The form data will be in JSON format.)
3. See the FlexiblePrice record in Django Admin. Appropriate fields should be filled, and it should have an FK to the course you created the form under and to the form submission. 

#### Any background context you want to provide?
Styling on the page will likely need some help - the template reuses the resource page template. We may want to do some additional styling to form fields too, and perhaps add defaults in for the form blocks. 

The CMS support for programs is not there in the system as of yet - that seemed like a bigger task so this PR doesn't address that. The code can be easily extended to support program pages once those are in (and it's marked where to do that). 

#### Screenshots (if appropriate)
Wagtail UI:
<img width="1401" alt="Form" src="https://user-images.githubusercontent.com/945611/167715013-cdba7302-7480-4006-9492-b7c48007ec2c.png">
<img width="1401" alt="Required Form Fields" src="https://user-images.githubusercontent.com/945611/167715011-f2e9679d-4e0f-4f2f-a18c-360647ec5632.png">

Form Display:
<img width="1401" alt="Basic Form" src="https://user-images.githubusercontent.com/945611/167715008-340339ec-2c1f-4609-a140-c81a94d22601.png">
<img width="1401" alt="Post-Submit Thank You" src="https://user-images.githubusercontent.com/945611/167715006-8fb0d238-70ca-4ac1-95bf-f7f461205b8a.png">

Submission Status:
<img width="1401" alt="Post-Submit Approved" src="https://user-images.githubusercontent.com/945611/167714998-e75e66e7-fd26-4ba5-8894-06c02b53038d.png">
<img width="1401" alt="Post-Submit Denied" src="https://user-images.githubusercontent.com/945611/167715004-db5db833-2fb7-4aa0-b44a-c7c8e2fe5822.png">
<img width="1401" alt="Post-Submit In progress" src="https://user-images.githubusercontent.com/945611/167715005-ca35ee33-308e-403d-9fb4-b60f0e108591.png">
